ddcommon/src/tag.rs

```diff
@@ -1,15 +1,68 @@
 // Copyright 2021-Present Datadog, Inc. https://www.datadoghq.com/
 // SPDX-License-Identifier: Apache-2.0
 
+use serde::{Deserialize, Serialize};
 use std::borrow::Cow;
 use std::fmt::{Debug, Display, Formatter};
 
-use serde::{Deserialize, Serialize};
+pub use static_assertions::{const_assert, const_assert_ne};
 
 #[derive(Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
 #[serde(transparent)]
 pub struct Tag {
-    value: String,
+    /// Many tags are made from literal strings, such as:
+    ///  - "language:native"
+    ///  - "src_library:libdatadog"
+    ///  - "type:timeout"
+    /// So being able to save allocations is nice.
+    value: Cow<'static, str>,
+}
+
+impl Tag {
+    /// Used by the `tag!` macro. Not meant to be used directly, please use
+    /// the macro instead.
+    /// # Safety
+    /// Do not use directly, use through the `tag!` macro which enforces the
+    /// safety invariants at compile time.
+    pub unsafe fn from_static_unchecked(value: &'static str) -> Self {
+        Self {
+            value: Cow::Borrowed(value),
+        }
+    }
+}
+
+// todo: what's a good way to keep these in-sync with Tag::from_value?
+// This can be a little more strict because it's compile-time evaluated.
+// https://docs.datadoghq.com/getting_started/tagging/#define-tags
+#[macro_export]
+macro_rules! tag {
```

Comment:
Put a doc comment explaining that this only works for compile time constant strings